### PR TITLE
Generalize contact rollups extraction from raw tables

### PR DESCRIPTION
**Pull request by bencodeorg**
[PLC-837](https://codedotorg.atlassian.net/secure/RapidBoard.jspa?rapidView=25&modal=detail&selectedIssue=PLC-837)

[reader's note: most substantial changes are in the `ContactRollupsRaw` model]

1. Does extraction of data from raw tables in raw SQL, which we found to be substantially (I forget relative difference, 15x?) faster in tests done on an ad hoc.
2. Makes extraction (somewhat) generalizable, in that it can be used to extract data from other tables. Current limitations:
-- Works with dashboard tables
-- Doesn't allow joins between tables
-- Assumes table has a column called `updated_at`
3. Puts a guard on full contact rollups process to skip Pardot steps by default.

 I haven't explicitly included parent email addresses as part of this PR, as I'd like to do that as a follow up. However, I have done most of the work that will be required to add parent email addresses or other fields we'd like to ingest. One remaining question is how to deal with sources (eg, `users`) where an email can appear in multiple rows (eg, `parent_email`) in a generalizable way.

## Testing story

I was able to manually run `bin/cron/build_contact_rollups_v2` on an adhoc pointed at a production clone -- these were the stats:

```
[2020-04-21 16:58:55 +0000] INFO: Truncate tables action completed without error in 1.02 seconds.
[2020-04-21 16:59:15 +0000] INFO: Extracts data from dashboard email_preferences action completed without error in 19.77 seconds.
[2020-04-21 17:04:42 +0000] INFO: Processes all extracted data action completed without error in 327.21 seconds.
[2020-04-21 17:05:01 +0000] INFO: Compares new and previously processed data action completed without error in 19.11 seconds.
[2020-04-21 17:05:16 +0000] INFO: Overwrites contact_rollups_final table action completed without error in 14.9 seconds.
```

I also added new tests to the `contact_rollups_raw` model, and updates `contact_rollups_processed` tests to cover more general behavior.

## Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: lib/cdo/contact_rollups/v2/contact_rollups.rb [hunk 1]
```diff
@@ -1,12 +1,12 @@
 class ContactRollupsV2
-  def self.build_contact_rollups(log_collector)
+  def self.build_contact_rollups(log_collector, pardot_steps=false)
```
Comment (molly-moen) [inline lib/cdo/contact_rollups/v2/contact_rollups.rb]: is the idea here that we are only going to do the full pardot steps very rarely, so we will do a code change to enable it? Nothing is calling this with true now right?

**Review by molly-moen — APPROVED** [review]
LGTM from what I know about contact rollups!

File: dashboard/app/models/contact_rollups_processed.rb [hunk 2]
```diff
@@ -83,17 +83,15 @@ def self.parse_contact_data(str)
...
   #   @see output of parse_contact_data method.
   # @return [Hash, nil] a hash containing opt_in key and value (could be nil)
   #   or nil if opt_in does not exist in the input.
-  def self.extract_opt_in(contact_data)
-    table = 'dashboard.email_preferences'
-    field = 'opt_in'
+  def self.extract_field(contact_data, table, field)
     return nil unless contact_data.key?(table) && contact_data[table].key?(field)
-    {opt_in: contact_data.dig(table, field)}
+    {field.to_sym => contact_data.dig(table, field)}
```
Comment (hacodeorg) [inline dashboard/app/models/contact_rollups_processed.rb:94]: I wonder if we should consistently keep key data type as string instead of symbol. The key comes from a JSON field in the database as a string, then after processing, will be written back to the database as a string.

File: lib/cdo/contact_rollups/v2/contact_rollups.rb (hunk 1 above)
Comment (bencodeorg) [reply lib/cdo/contact_rollups/v2/contact_rollups.rb]: @hacodeorg may object -- was just a convenience way for me to run the non-Pardot steps. I think it might be a good idea generally though to have to explicit about running the Pardot steps -- we both have Pardot API keys in our locals.yml, so it doesn't seem implausible that we might accidentally start writing to Pardot. Some of the data in Pardot didn't come from our production DB and has no backup as far as I know, so I'm in favor of putting guardrails up.

File: lib/cdo/contact_rollups/v2/contact_rollups.rb [hunk 3]
```diff
@@ -18,14 +18,16 @@ def self.build_contact_rollups(log_collector)
...
 
-    log_collector.time!('Downloads new email-Pardot ID mappings') do
-      ContactRollupsPardotMemory.add_and_update_pardot_ids
-    end
-    log_collector.time!('Creates new Pardot prospects') do
-      ContactRollupsPardotMemory.create_new_pardot_prospects
-    end
-    log_collector.time!('Updates existing Pardot prospects') do
-      ContactRollupsPardotMemory.update_pardot_prospects
+    if pardot_steps
```
Comment (hacodeorg) [inline lib/cdo/contact_rollups/v2/contact_rollups.rb]: I think this belongs to another PR when we want to turn on the pipeline (with new cronjob entry). However, this is a 2-line change so it's okay if you want to keep it here.
```suggestion
if sync_with_pardot
```

File: dashboard/app/models/contact_rollups_raw.rb [hunk 4]
```diff
@@ -22,22 +22,38 @@ def self.truncate_table
...
+        #{extract_columns_into_mysql_json(data_columns)} AS data,
+        #{source_table}.updated_at AS data_updated_at,
+        NOW() AS created_at,
+        NOW() AS updated_at
+      FROM #{source_table}
+      WHERE #{email_column} IS NOT NULL AND #{email_column} != ''
+    SQL
+  end
+
+  # @param columns [Array] Column names to reshape
```
Comment (hacodeorg) [inline dashboard/app/models/contact_rollups_raw.rb:54]: Could you and an @example here.

File: dashboard/app/models/contact_rollups_raw.rb [hunk 5]
```diff
@@ -22,22 +22,38 @@ def self.truncate_table
...
+  # @return [String] MySQL JSON_OBJECT() syntax for insertion
+  def self.extract_columns_into_mysql_json(columns)
+    return 'NULL' if columns.empty?
 
-      # ben to do: handle if required fields are missing, or if there's a duplicate?
-      # currently will fail if any required fields are missing
-      import! raw_contacts
-    end
+    last = columns.pop
+    'JSON_OBJECT(' + columns.map {|column| "'#{column}', #{column}, "}.join + "'#{last}', #{last}" + ')'
```
Comment (hacodeorg) [inline dashboard/app/models/contact_rollups_raw.rb]: You can do this without using `last`
```suggestion
'JSON_OBJECT(' + columns.map {|col| "'#{col}',#{col}"}.join(',') + ')'
```

File: dashboard/app/models/contact_rollups_raw.rb [hunk 6]
```diff
@@ -22,22 +22,38 @@ def self.truncate_table
...
+      INSERT INTO #{ContactRollupsRaw.table_name} (email, sources, data, data_updated_at, created_at, updated_at)
+      SELECT
+        #{email_column},
+        'dashboard.#{source_table}' AS sources,
+        #{extract_columns_into_mysql_json(data_columns)} AS data,
+        #{source_table}.updated_at AS data_updated_at,
+        NOW() AS created_at,
+        NOW() AS updated_at
+      FROM #{source_table}
+      WHERE #{email_column} IS NOT NULL AND #{email_column} != ''
```
Comment (hacodeorg) [inline dashboard/app/models/contact_rollups_raw.rb:45]: +1 for checking this

File: dashboard/app/models/contact_rollups_raw.rb [hunk 7]
```diff
@@ -22,22 +22,38 @@ def self.truncate_table
...
-        raw_contact = {
-          email: email_preference.email,
-          sources: "dashboard.#{EmailPreference.table_name}",
-          data: {opt_in: email_preference.opt_in},
-          data_updated_at: email_preference.updated_at
-        }
-        raw_contacts << raw_contact
-      end
+  # Generates insert query to run, and runs that query once constructed.
+  def self.extract_from_table(source_table, data_columns, email_column)
```
Comment (hacodeorg) [inline dashboard/app/models/contact_rollups_raw.rb]: A slightly different approach to consider:
+ Keeps the original function signature `extract_email_preferences`, no parameters. 
It is easier for callers (ContactRollupsV2 and ContactRollupsRawTest) to invoke `ContactRollupsRaw.extract_email_preferences` without worrying about any parameters used internally by the function such as source table, email column and data columns.

+ `extract_from_table_query` can take a source table or a source query to construct an insert statement. 
If data comes from a permanent table, the result will be `INSERT INTO ... SELECT ... FROM <a table>`. If data comes from another query, the result will be `INSERT INTO ... SELECT ... FROM (<a query>) AS inner_query`. This will solve the problem of prepping data (such as removing duplicate emails, changing column names, and joining tables) before ingress.

File: dashboard/test/models/contact_rollups_processed_test.rb [hunk 8]
```diff
@@ -54,24 +54,27 @@ class ContactRollupsProcessedTest < ActiveSupport::TestCase
...
+
     test_cases = [
-      {input: {}, expected_output: nil},
-      {input: {'dashboard.email_preferences' => {}}, expected_output: nil},
-      {input: {'dashboard.another_table' => {opt_in: 1}}, expected_output: nil},
-      {input: {'dashboard.email_preferences' => {'opt_in' => 0}}, expected_output: {opt_in: 0}},
-      {input: {'dashboard.email_preferences' => {'opt_in' => 1}}, expected_output: {opt_in: 1}},
-      {input: {'dashboard.email_preferences' => {'opt_in' => nil}}, expected_output: {opt_in: nil}}
+      {input: [{}, nil, nil], expected_output: nil},
+      {input: [{'dashboard.email_preferences' => {}}, table_name, field_name], expected_output: nil},
```
Comment (hacodeorg) [inline dashboard/test/models/contact_rollups_processed_test.rb]: Replace all references to `dashboard.email_preferences` by `table_name`, and replace `opt_in` by `field_name`. You can even omit the `_name` suffix.

File: dashboard/test/models/contact_rollups_raw_test.rb [hunk 9]
```diff
@@ -3,9 +3,9 @@
 class ContactRollupsRawTest < ActiveSupport::TestCase
   test 'extract_email_preferences creates records as we would expect' do
     email_preference = create :email_preference
-    ContactRollupsRaw.extract_email_preferences
+    ContactRollupsRaw.extract_from_table('email_preferences', ['opt_in'], 'email')
 
-    expected_data = {opt_in: email_preference.opt_in}
+    expected_data = {opt_in: email_preference.opt_in ? 1 : 0}
```
Comment (hacodeorg) [inline dashboard/test/models/contact_rollups_raw_test.rb:9]: Why we need this value conversion?

File: dashboard/app/models/contact_rollups_raw.rb [hunk 10]
```diff
@@ -22,22 +22,38 @@ def self.truncate_table
...
+        NOW() AS created_at,
+        NOW() AS updated_at
+      FROM #{source_table}
+      WHERE #{email_column} IS NOT NULL AND #{email_column} != ''
+    SQL
+  end
+
+  # @param columns [Array] Column names to reshape
+  # @return [String] MySQL JSON_OBJECT() syntax for insertion
+  def self.extract_columns_into_mysql_json(columns)
```
Comment (hacodeorg) [inline dashboard/app/models/contact_rollups_raw.rb]: How about just `create_json_object` + an 1-line explanation in method comment?

File: dashboard/test/models/contact_rollups_raw_test.rb (hunk 9 above)
Comment (hacodeorg) [reply dashboard/test/models/contact_rollups_raw_test.rb:9]: Update: MySql store boolean as 0/1 while Rails uses true/false.

**Review by hacodeorg — APPROVED** [review]
Looks great! Thanks, Ben! 
There is only 1 non-blocking question left about consistently using 1 type of hash key (string vs. symbol).

File: dashboard/app/models/contact_rollups_processed.rb (hunk 2 above)
Comment (bencodeorg) [reply dashboard/app/models/contact_rollups_processed.rb:94]: I think that's a good idea -- that said, I didn't want to get into making that change in this PR, but something we can update going forward.
